SU2_CFD/include/limiters/CLimiterDetails.hpp

```diff
@@ -0,0 +1,335 @@
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2019, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+/*!
+ * \brief A traits class for limiters, see notes for "computeLimiters_impl()".
+ * \note There is no default implementation (the code will compile but not
+ *       link) specialization is mandatory.
+ */
+template<ENUM_LIMITER LimiterKind>
+struct CLimiterDetails
+{
+  /*!
+   * \brief Compute any global value that may be needed by the other functions.
+   * \note This function is called once by multiple threads.
+   */
+  template<class FieldType>
+  inline void preprocess(CGeometry&, CConfig&, size_t varBegin,
+                         size_t varEnd, const FieldType&);
+
+  /*!
+   * \brief Geometric modifier (e.g. increase limiting near sharp edges).
+   * \note This function is called once per point inside an AD pre-
+   *       -accumulation region, newly used variables should be registered.
+   */
+  inline su2double geometricFactor(size_t iPoint, CGeometry&) const;
+
+  /*!
+   * \brief Smooth (usually) function of the maximum/minimum (positive/negative)
+   *        gradient projections onto the edges, and the deltas over direct neighbors.
+   *        Both proj and delta may be 0.0, beware of divisions.
+   * \note This function is called twice (min/max) per point per variable
+   *       (also inside an AD pre-accumulation region).
+   */
+  inline su2double limiterFunction(size_t iVar, su2double proj, su2double delta) const;
+};
+
+
+/*!
+ * \brief Common small functions used by limiters.
+ */
+namespace LimiterHelpers
+{
+  inline passivedouble epsilon() {return std::numeric_limits<passivedouble>::epsilon();}
+
+  inline su2double venkatFunction(su2double proj, su2double delta, su2double eps2)
+  {
+    su2double y = delta*(delta+proj) + eps2;
+    return (y + delta*proj) / (y + 2*proj*proj);
+  }
+
+  inline su2double raisedSine(su2double dist)
+  {
+    su2double factor = 0.5*(1.0+dist+sin(PI_NUMBER*dist)/PI_NUMBER);
+    return max(0.0, min(factor, 1.0));
+  }
+}
+
+
+/*!
+ * \brief Barth-Jespersen specialization.
+ */
+template<>
+struct CLimiterDetails<BARTH_JESPERSEN>
+{
+  su2double eps2;
+
+  /*!
+   * \brief Set a small epsilon to avoid divisions by 0.
+   */
+  template<class... Ts>
+  inline void preprocess(Ts&...) {eps2 = LimiterHelpers::epsilon();}
```

Comment:
Had to look up what the `...` was for. Didn't know about parameter packs, very cool. 
Also super clean implementation of limiters.

Reply:
Thank you for the review @jayantmukho, to be honest I'm no specialist either, I am only using them to make those methods forward compatible in case some of the specializations need to be passed extra parameters in the future.